In [1]:
from sklearn import datasets
from sklearn.cluster import AgglomerativeClustering  #Hierarchical Clustering which is implemented in scikit-learn as AgglomerativeClustering
import sklearn.metrics as sm
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

iris = datasets.load_iris()

x = iris.data
y = iris.target

In [2]:
#Predicting number of clusters using Silhouette
range_n_clusters = [2, 3, 4, 5, 6]
for n_clusters in range_n_clusters:
	clusterer = AgglomerativeClustering(n_clusters=n_clusters, affinity='euclidean',linkage='ward')
	cluster_labels = clusterer.fit_predict(x)
	silhouette_avg = sm.silhouette_score(x, cluster_labels)
	print("For n_clusters =", n_clusters,"The average silhouette_score is :", silhouette_avg)

('For n_clusters =', 2, 'The average silhouette_score is :', 0.68639305432850328)
('For n_clusters =', 3, 'The average silhouette_score is :', 0.55409729079614556)
('For n_clusters =', 4, 'The average silhouette_score is :', 0.48873984907752543)
('For n_clusters =', 5, 'The average silhouette_score is :', 0.48415535260111559)
('For n_clusters =', 6, 'The average silhouette_score is :', 0.36188451501872154)


In [3]:
#Hierarchical Clustering which is implemented in scikit-learn as AgglomerativeClustering
model = AgglomerativeClustering(n_clusters=3, affinity='euclidean', linkage='ward')
model.fit(x)

predY = np.choose(model.labels_, [1, 0, 2]).astype(np.int64)
print (model.labels_)
print (predY)
print iris.target

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 2 2 2 2 0 2 2 2 2
 2 2 0 0 2 2 2 2 0 2 0 2 0 2 2 0 0 2 2 2 2 2 0 0 2 2 2 0 2 2 2 0 2 2 2 0 2
 2 0]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 2 2 2 2 1 2 2 2 2
 2 2 1 1 2 2 2 2 1 2 1 2 1 2 2 1 1 2 2 2 2 2 1 1 2 2 2 1 2 2 2 1 2 2 2 1 2
 2 1]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


In [4]:
# Performance Metrics
print sm.accuracy_score(y, predY)

	
# Confusion Matrix
print sm.confusion_matrix(y, predY)

0.893333333333
[[50  0  0]
 [ 0 49  1]
 [ 0 15 35]]


# Discussion

Here also we got a number of cluster value from Silhoutte as 2. This is not giving enough accuracy. So, we choose 3 as in Kmeans. The accuracy is seen as 89%, which is quite good.